src/System.Security.Cryptography.Pkcs/src/Internal/Cryptography/Pal/Unix/PkcsPalOpenSsl.Encrypt.cs

```diff
@@ -70,14 +71,94 @@ internal sealed partial class PkcsPalOpenSsl : PkcsPal
+
+            byte[][] contentType = originalMessage.ReadAndSplitNextEncodedValue();
+            DerSequenceReader envelopedDataReader = originalMessage.ReadExplicitContextSequence();
+
+            byte[][] version = envelopedDataReader.ReadAndSplitNextEncodedValue();
+
+            const byte ConstructedContextSpecificTag = DerSequenceReader.ContextSpecificTagFlag | DerSequenceReader.ConstructedFlag;
+
+            byte[][] originatorInfo = (envelopedDataReader.PeekTag() == ConstructedContextSpecificTag) ?
+                originatorInfo = envelopedDataReader.ReadAndSplitNextEncodedValue() :
+                new byte[][] { Array.Empty<byte>(), Array.Empty<byte>(), Array.Empty<byte>() };
+
+            byte[][] recipientInfos = envelopedDataReader.ReadAndSplitNextEncodedValue();
+            byte[][] encryptedContentInfo = envelopedDataReader.ReadAndSplitNextEncodedValue();
+
+            byte[][] envelopedData = DerEncoder.ConstructSegmentedExplicitSequence(
+                0 /* Context number */,
+                version,
+                originatorInfo,
+                recipientInfos,
+                encryptedContentInfo,
+                EncodeAttributes(unprotectedAttributes));
+
+            return DerEncoder.ConstructSequence(
+                contentType,
+                envelopedData);
+        }
+
+        private byte[][] EncodeAttributes(CryptographicAttributeObjectCollection unprotectedAttributes)
+        {
+            // In EnvelopedData UnprotectedAttributes are included as:
+            //
+            // unprotectedAttrs[1] IMPLICIT UnprotectedAttributes OPTIONAL
+            //
+            // UnprotectedAttributes ::= SET SIZE (1..MAX) OF Attribute
+            // Attribute::= SEQUENCE {
+            //      attrType OBJECT IDENTIFIER,
+            //      attrValues SET OF AttributeValue }
+
+            byte[][][] setOfAttrs = new byte[unprotectedAttributes.Count][][];
+            
+            for (int i = 0; i < unprotectedAttributes.Count; i++)
+            {
+                byte[][] attrType = DerEncoder.SegmentedEncodeOid(unprotectedAttributes[i].Oid);
+                byte[][][] attrValues = new byte[unprotectedAttributes[i].Values.Count][][];
+
+                for (int j = 0; j < attrValues.Length; j++)
+                {
+                    attrValues[j] = DerSequenceReader.SplitValue(unprotectedAttributes[i].Values[j].RawData);
+                }
+
+                byte[][] attrValueSet = DerEncoder.ConstructSegmentedSet(attrValues);
+                byte[][] encodedAttribute = DerEncoder.ConstructSegmentedSequence(attrType, attrValueSet);
+                setOfAttrs[i] = encodedAttribute;
+            }
+
+            // There's no difference between an implicit set and an implicit sequence in the tag,
+            // so we can just call this.
+            return DerEncoder.ConstructSegmentedImplicitSequence(1 /* Context number */, setOfAttrs);
```

Comment:
Seems like that's a decision for the encoder to make, rather than a comment here.